build/source/engine/check_icond.f90

```diff
@@ -0,0 +1,292 @@
+   mLayerLayerType   => indxData%gru(iGRU)%hru(iHRU)%var(iLookINDEX%layerType)%dat            , & ! type of layer (ix_soil or ix_snow)
+   ! model parameters
+   vGn_alpha         => mparData%gru(iGRU)%hru(iHRU)%var(iLookPARAM%vGn_alpha)                , & ! van Genutchen "alpha" parameter (m-1)
+   vGn_n             => mparData%gru(iGRU)%hru(iHRU)%var(iLookPARAM%vGn_n)                    , & ! van Genutchen "n" parameter (-)
+   theta_sat         => mparData%gru(iGRU)%hru(iHRU)%var(iLookPARAM%theta_sat)                , & ! soil porosity (-)
+   theta_res         => mparData%gru(iGRU)%hru(iHRU)%var(iLookPARAM%theta_res)                , & ! soil residual volumetric water content (-)
+   snowfrz_scale     => mparData%gru(iGRU)%hru(iHRU)%var(iLookPARAM%snowfrz_scale)            , & ! scaling parameter for the snow freezing curve (K-1)
+   FCapil            => mparData%gru(iGRU)%hru(iHRU)%var(iLookPARAM%FCapil)                     & ! fraction of pore space in tension storage (-)
+   )  ! (associate local variables with model parameters)
+
+   ! compute the maximum volumetric fraction of liquid water -- used to avoid problems of super-saturation in the moisture-based form of Richards' equation
+   maxVolFracLiq = theta_sat - 1.e-4_dp
+
+   ! compute the van Genutchen "m" parameter (-)
+   vGn_m = 1._dp - 1._dp/vGn_n
+
+   ! compute the constant in the freezing curve function (m K-1)
+   kappa  = (iden_ice/iden_water)*(LH_fus/(gravity*Tfreeze))  ! NOTE: J = kg m2 s-2
+
+   ! modify the liquid water and ice in the canopy
+   if(scalarCanopyIce > 0._dp .and. scalarCanopyTemp > Tfreeze)then
+    message=trim(message)//'canopy ice > 0 when canopy temperature > Tfreeze'
+    err=20; return
+   end if
+   fLiq = fracliquid(scalarCanopyTemp,snowfrz_scale)  ! fraction of liquid water (-)
+   tWat = scalarCanopyLiq + scalarCanopyIce           ! total water (kg m-2)
+   scalarCanopyLiq = fLiq*tWat                        ! mass of liquid water on the canopy (kg m-2)
+   scalarCanopyIce = (1._dp - fLiq)*tWat              ! mass of ice on the canopy (kg m-2)
+
+   ! number of layers
+   nLayers = gru_struc(iGRU)%hruInfo(iHRU)%nSnow + gru_struc(iGRU)%hruInfo(iHRU)%nSoil
+   nSnow   = gru_struc(iGRU)%hruInfo(iHRU)%nSnow
+
+   ! loop through all layers
+   do iLayer=1,nLayers
+
+    ! compute liquid water equivalent of total water (liquid plus ice)
+    if (iLayer>nSnow) then ! soil layer = no volume expansion
+     scalarTheta = mLayerVolFracIce(iLayer) + mLayerVolFracLiq(iLayer)
+    else ! snow layer = volume expansion allowed
+     scalarTheta = mLayerVolFracIce(iLayer)*(iden_ice/iden_water) + mLayerVolFracLiq(iLayer)
+    end if
+
+    ! check that the initial volumetric fraction of liquid water and ice is reasonable
+    select case(mlayerLayerType(iLayer))
+
+     ! ***** snow
+     case(ix_snow)
+      ! (check liquid water)
+      if(mLayerVolFracLiq(iLayer) < 0._dp .or. mLayerVolFracLiq(iLayer) > 1._dp)then
```

Comment:
Can the checks be separated like:
```
if (mLayerVolFracLiq(iLayer) < 0._dp) less than Zero error
if (mLayerVolFracLiq(iLayer) > 1._dp) larger than One error
```
And `mLayerVolFracLiq(iLayer)` will be checked with `theta_res` and `theta_sat` so is it a redundant check here for `[0, 1]`?